DSC550-T302<br>
Week 3 Assignment<br>
Jeremy Barton

## Preparing Data for a Custom Model, "Bag of Words" Method

#### Introduction

This notebook will classify movies review as positive or neative assuming a polarity score of >= 0 as positive, and <= 0 as negative The bag of words model will then be compared to random guessing to check it's accuracy.

1) Importing the movie review data as a data frame and ensuring that the data is loaded properly.

In [34]:
import pandas as pd
from textblob import TextBlob
from sklearn.metrics import accuracy_score

# using sep='\t' tells pandas this is a 
#   tab separated document
movie_reviews = pd.read_csv("labeledTrainData.tsv", sep='\t')

Check the data exists.

In [35]:
movie_reviews.head(5)

,id,sentiment,review
0,5814_8,1,With all this stuff going down at the moment w...
1,2381_9,1,"\The Classic War of the Worlds\"" by Timothy Hi..."
2,7759_3,0,The film starts with a manager (Nicholas Bell)...
3,3630_4,0,It must be assumed that those who praised this...
4,9495_8,1,Superbly trashy and wondrously unpretentious 8...


In [36]:
movie_reviews.shape

(25000, 3)

2) How many of each positive and negative reviews are there?

To calculate this the 1s and 0s should be totalled up. The function to use for this is just `.value_counts()`.

The below line sumes the 1s and 0s of the sentiment column from the movie_reviews dataset.

<b>Positive</b>

In [37]:
# .int() turns the result into an integeer
# .sum() is used on the sentiment indicator "1"
#   from the "movie_reviews" dataframe
positive = int((movie_reviews['sentiment'] == 1).sum())
positive

12500

<b>Negative</b>

In [38]:
negative = int((movie_reviews['sentiment'] == 0).sum())
negative

12500

3) Use TextBlob to classify each movie review as positive or negative. Assume that a polarity score greater than or equal to zero is a positive sentiment and less than 0 is a negative sentiment.

In [39]:
# create analysis container
analysis_df = []

# iterate through movie reviews
#   and collect text 
for text in movie_reviews['review']:
    # calculate polarity score using TextBlob
    pol = TextBlob(text).sentiment.polarity
    # use TextBlob's build in .sentiment.polarity
    #   core usage methods

    # checking if the polarity is greater than
    # or equal to zero
    if pol >=0:
        sentiment = 1 # marks a positive review
    else:
        sentiment = 0 # marks a negative review
    
    # append the results to a list
    analysis_df.append({
        'review': text,
        'polarity': pol,
        'predicted': sentiment
    })

After that's finished running, the results can now be printed.

In [40]:
# convert to dataframe and add colums
#   to compare actual and predicted results
analysis_df = pd.DataFrame(analysis_df)
analysis_df['actual'] = movie_reviews['sentiment'].values

print(analysis_df.head())

                                              review  polarity  predicted  \
0  With all this stuff going down at the moment w...  0.001277          1   
1  \The Classic War of the Worlds\" by Timothy Hi...  0.256349          1   
2  The film starts with a manager (Nicholas Bell)... -0.053941          0   
3  It must be assumed that those who praised this...  0.134753          1   
4  Superbly trashy and wondrously unpretentious 8... -0.024842          0   

   actual  
0       1  
1       1  
2       0  
3       0  
4       1  


In the dataframe, the "predicted" represents what TextBlob calculated the sentiment was. The "actual" column is what the actual sentiment was according to the data.

4) Check the accuracy of this model. Is this model better than random guessing?

Checking the accuracy of the results can be performed using <b>scikit-learn's</b> `accuracy_score`. It compares two columns by taking the true and predicted labels, counts how many `predicted` values match the `actual` values and divides the total number of rows.

In [44]:
# accuracy_score() takes two parameters
#   the actual results and the predicted values
accuracy = accuracy_score(analysis_df['actual'], analysis_df['predicted'])
print(f"Accuracy: {accuracy:.2%}")

Accuracy: 68.52%


So with a total of 25,000 total reviews, a 68.52% accuracy this means 17,130 reviews were correctly classified, and 7,870 were incorrectly classified.

<b>What are some issues with this test?</b>

TextBlob is a general purpose sentiment tool, which means it isn't traied for movie reviews specifically. As typically plauged with in natural language processing these movie reviews may include, sarcasm, nuance and complex that simpler polarity scoring techniques might miss.

<b>Is this better than random guessing?</b>

Considering this provides a <b>quantitative</b> answer to the question, it certainly provides a <i>different</i> approach to predicting sentiment.

Assuming that guessing involves simply reading the text and making a guess on the review being positive or negative and due to the shere amount of sarcasm, nuance and inside jokes that could appear in movie reviews, random guessing in this scenario may be more useful.

Using this level of text analysis on movie reviews may only supplemental to a guess-based approach.